File: DotNetPractices/ProblemSolving/Codility/QuickSort.cs
class QuickSort
{
    void Swap(ref int a, ref int b)
    {
        int t = a;
        a = b;
        b = t;
    }

    int Partition(int[] a, int low, int high)
    {
        int pivot = a[high];
        int l = low;

        for (int i = low; i < high - 1; i++)
        {
            if (a[i] < pivot)
            {
                Swap(ref a[i], ref a[l]);
                l++;
            }
        }
        Swap(ref a[l], ref a[high]);
        return l;
    }


    public void Sort(int[] A, int low, int high)
    {
        while (low < high)
        {
            int pi = Partition(A, low, high);

            Sort(A, low, pi - 1);
            Sort(A, pi + 1, high);
        }
    }
}

File: DotNetPractices/ProblemSolving/Cpp/algos/quick_sort.cpp
#include <iostream>
#include <vector>
using namespace std;

void swap(int* a, int* b) {
    int t = *a;
    *a = *b;
    *b = t;
}

int partition(vector<int>& v, int l, int h) {
    int pivot = v[h];
    int i = l;

    for (int j = l; j < h; j++)
    {
        if (v[j] < pivot) swap(&v[i++], &v[j]);
    }

    swap(&v[i], &v[h]);

    return i;
}

void sort(vector<int>& v, int l, int h) {
    if (l < h) {
        int p = partition(v, l, h);
        sort(v, l, p - 1);
        sort(v, p + 1, h);
    }
}


void print(vector<int> v) {
    for (int i = 0; i < v.size(); i++)
        cout << v[i] << " ";
}


int main()
{
    string s;

    vector<int> v;
    v.push_back(5);
    v.push_back(2);
    v.push_back(3);
    v.push_back(6);
    v.push_back(1);
    v.push_back(2);
    v.push_back(4);

    sort(v, 0, v.size() - 1);
    print(v);

    return 0;
}
